Rubberduck.Refactorings/ExtractInterface/ExtractInterfaceBaseRefactoring.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Linq;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Refactorings.AddInterfaceImplementations;
+using Rubberduck.VBEditor.SafeComWrappers;
+
+namespace Rubberduck.Refactorings.ExtractInterface
+{
+    public class ExtractInterfaceBaseRefactoring : BaseRefactoringWithSuspensionBase<ExtractInterfaceModel>
+    {
+        private readonly ICodeOnlyBaseRefactoring<AddInterfaceImplementationsModel> _addImplementationsRefactoring;
+        private readonly IDeclarationFinderProvider _declarationFinderProvider;
+
+        public ExtractInterfaceBaseRefactoring(
+            AddInterFaceImplementationsBaseRefactoring addImplementationsRefactoring,
+            IDeclarationFinderProvider declarationFinderProvider,
+            IParseManager parseManager, 
+            IRewritingManager rewritingManager) 
+            : base(parseManager, rewritingManager)
+        {
+            _addImplementationsRefactoring = addImplementationsRefactoring;
+            _declarationFinderProvider = declarationFinderProvider;
+        }
+
+        protected override bool RequiresSuspension(ExtractInterfaceModel model)
+        {
+            return true;
+        }
+
+        protected override void Refactor(ExtractInterfaceModel model, IRewriteSession rewriteSession)
+        {
+            AddInterface(model, rewriteSession);
+        }
+
+        private void AddInterface(ExtractInterfaceModel model, IRewriteSession rewriteSession)
+        {
+            var targetProject = model.TargetDeclaration.Project;
+            if (targetProject == null)
+            {
+                return; //The target project is not available.
+            }
+
+            AddInterfaceClass(model.TargetDeclaration, model.InterfaceName, GetInterfaceModuleBody(model));
+
+            var rewriter = rewriteSession.CheckOutModuleRewriter(model.TargetDeclaration.QualifiedModuleName);
+
+            var firstNonFieldMember = _declarationFinderProvider.DeclarationFinder.Members(model.TargetDeclaration)
+                                            .OrderBy(o => o.Selection)
+                                            .First(m => ExtractInterfaceModel.MemberTypes.Contains(m.DeclarationType));
+            rewriter.InsertBefore(firstNonFieldMember.Context.Start.TokenIndex, $"Implements {model.InterfaceName}{Environment.NewLine}{Environment.NewLine}");
+
+            AddInterfaceMembersToClass(model, rewriteSession);
+        }
+
+        private void AddInterfaceClass(Declaration implementingClass, string interfaceName, string interfaceBody)
+        {
```

Comment:
This PR is really great.  Looking ahead a little bit...the `MoveMember` refactoring will also create new modules.  I'm wondering if `BaseRefactoringBase` could/should provide a default 'add new module' capability.  Looks like adding a `ComponentType` parameter would be all that is needed to generalize it.

Reply:
No, it is not a concern of the refactoring action base class to provide a method to add components. That just forces more dependencies on refactorings that don't necessarily need the capability.

Instead, just constructor inject an `IAddComponentService` into the refactoring action that needs to add components.

Reply:
It you look at the latest state of this PR, you can see the version with the add component service.